### PR TITLE
macros: addition of `badge` macro

### PR DESCRIPTION
**Pull request by JavierDelgadoFernandez**
* Adds a macro that generates a badge using a title, a value and
  optionally, a background color.

Signed-off-by: Javier Delgado <javier.delgado.fernandez@cern.ch>

### CODE REVIEW

**Comment by JavierDelgadoFernandez** [comment]
@lnielsen Instead of using a link to a external service, I am just generating a SVG. It is faster and in this way there is no need to depend on third-parties services.

I was checking and it is easy to create a ``to_png`` filter receiving a SVG. And creating another macro, you can just pass the result of the call to the SVG generation to a this filter and create an ```` node with the image embedded using base64. However, I think it is better to force the user to use SVG...

**Comment by JavierDelgadoFernandez** [comment]
Example:
badge

**Comment by lnielsen** [comment]
:tada: Great news! I think we'll need to add utility functions that takes care of generating the SVG (probably by rendering a template that is currently in your macro). Also, then you can make another utility function that makes the PNG (probably as a BytesIO stream or similar).

The template macro will then just use these utility functions to make the badge. The reason for utility functions are that I need to plug them on endpoints eg. ``/githubrepo/badge.svg``, where a template macro won't work.

**Comment by lnielsen** [comment]
btw, PNG is only for legacy support.

**Comment by lnielsen** [comment]
Last extra think: If you try out the badge on zenodo, you can usually click the badge and grab some template code to embed the badge - a extra template macro that can do this would also be great!

File: invenio_formatter/templates/invenio_formatter/macros/meta.html [hunk 1]
```diff
@@ -73,3 +73,27 @@
...
+        <mask id="a">
+            <rect width="86" height="20" rx="3" fill="#fff"/>
+        </mask>
+        <g mask="url(#a)">
+            <path fill="#555" d="M0 0h31v20H0z"/>
+            <path fill="{{ color }}" d="M31 0h55v20H31z"/>
+            <path fill="url(#b)" d="M0 0h86v20H0z"/>
+        </g>
+        <g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11">
+            <text x="15.5" y="15" fill="#010101" fill-opacity=".3">{{ title }}</text>
```
Comment (JavierDelgadoFernandez) [inline invenio_formatter/templates/invenio_formatter/macros/meta.html]: Ok, I have seen that these values for the X and Y are fixed but they should depend on the size of the title and value.

**Comment by JavierDelgadoFernandez** [comment]
@lnielsen There is a problem with the accuracy of the box size. Since it is not using mono spaced fonts, it is not easy to calculate the proper value. In Shields.IO, they are using pdfkit, which is a Ruby project.

**Comment by jirikuncar** [comment]
> There is a problem with the accuracy of the box size.

@JavierDelgadoFernandez Have you tried
https://pillow.readthedocs.org/en/3.2.x/reference/ImageDraw.html#PIL.ImageDraw.PIL.ImageDraw.Draw.textsize ?

**Comment by JavierDelgadoFernandez** [comment]
@jirikuncar Yep, I am testing it now again and it looks to be working properly. I was checking yesterday several libraries (Pygame (Too much dependencies), ttfquery(Too outdated, not working with Py3)) and this one as well, but I gave up because the dependencies (``libtiff5-dev libjpeg8-dev zlib1g-dev \
    libfreetype6-dev liblcms2-dev libwebp-dev tcl8.6-dev tk8.6-dev python-tk``).

**Comment by JavierDelgadoFernandez** [comment]
@lnielsen It is ready from my side.

File: invenio_formatter/context_processors/badges.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def get_text_length(*args):
+    """Measure the size of string rendered with a TTF no-nomospaced fonts."""
+    txt = Image.new('RGBA', (16, 16), (255, 255, 255, 0))
+    d = ImageDraw.Draw(txt)
+    font = ImageFont.truetype("DejaVuSans", 11)
+    result = ()
+    for value in args:
+        result = result + (d.textsize(value, font=font)[0], )
```
Comment (jirikuncar) [inline invenio_formatter/context_processors/badges.py]: :+1:

**Pull request merged**